### DISCUSSION TITLE
elastoplasticity based on cable plugin

### DISCUSSION THREAD

**Discussion by mathl79** [google-deepmind/mujoco#549] [Ideas]
Hi,

I'm trying to use MuJoCo for bending process simulation of metal wires and rods.
Therefore i want to start working on an elasto**plastic** plugin based on the new the elastic cable plugin to implement plastic deformations.

My idea of a very simplified plastic deformation cable model is to add bending & twisting angle limit values to the material parameters, that will be used to conditionally change the initial omega values towards the actual omega values of the model. 

Is this approach of altering the initial omega values at simulation time possible ?

The implementation of nonlinear material parameters (stress-strain table) would be the next step then...

**Comment by quagla** [comment]
Hi,

The idea of elastoplastic / viscoplastic rods is very interesting! 

I believe the most principled approach would be to use _strain_ limits rather than _angle_ limits because in the case of nonlinear deformations it's not possible (or very hard) to separate twisting and bending deformations by looking at angles, whereas it's automatic in the strains. As you are suggesting, one possibility is indeed to update `omega0` inside the plugin at every call of `Compute`, which would effectively become a new state hidden from the rest of MuJoCo. I think this is fine if you have some quasistatic assumption, i.e. that the plastic strains are generated more rapidly that the rest of the dynamics. Definitely simple and worth trying!

If you instead need to more properly integrate the new states, a little more work might be needed, but it's also doable. Are you thinking about rate-dependent or independent plasticity?

**Reply by mathl79** [reply]
thanks for your quick and your positive feedback!

At my use case, the quasi static / permanent geometric deformation is more important than the dynamic behavior. 
Therefore the plastic domain does not need to be rate dependent. The Energy loss due to friction would be rate dependent and therefore the stiffness would be decreased due to temperature rise, but this is not in my scope :-)
I think dynamic plasticity or viscosity can be assumed to be equivalent to damping, right?

But, if the calculation is based on strains values, it would be possible to change the stiffness dynamically from a static stress-strain lookup table 
the omega value updates representing plastic deformation are then calculated based on the change between the initial stiffness E0 and the strain-dependent stiffness E(strain) value of the material parameters.

**Reply by quagla** [reply]
Yes, viscosity and damping are basically the same thing and in general all forces (elastic, plastic, viscous) look similar in the sense that they penalize a difference `omega-omega0`, but the meaning of omega0 changes depending on the force. For example, it is common to use omega0 at the last time step for viscosity/damping.

Yes, it's possible to have a stiffness lookup table (are you thinking about hardening?) although you have to compute the stress, which depends on the stiffness itself, as in `Compute` you compute force from displacement. So the workflow would be displacement->total strain->subtract total-plastic->stiffness->stress->update plastic strain. Or do you have another idea for computing the stiffness before the stress?

**Reply by mathl79** [reply]
![image](https://user-images.githubusercontent.com/88651276/198285652-260206d9-3655-43ed-a48d-94f01949d5a1.png)

Yes, I think of having a detailed material representation including strain hardening, except material destruction.
The stress and strain is fully defined with this kind of lookup table / diagram.
Some kind of interpolation is necessary to get stress from any strain point. 

I am not sure what the best approach would be. 
starting from strain would lead to this workflow:
strain -> stiffness E from LUT (local derivative)  -> calculate new strain from linear backward projection parallel to elastic portion
![image](https://user-images.githubusercontent.com/88651276/198293167-845cd2b3-b866-403d-a397-5c9963bc9c3f.png)

does this make sense to you ?

**Reply by quagla** [reply]
I would personally start with a simple Von Mises approach as a proof-of-concept. Something like: given the strain, compute the stress as if the deformation was purely elastic, then get the plastic strain by rescaling the total strain with the relative difference between the stress and the yield stress. Finally, get the true stress using the difference between the total strain and the plastic strain multiplied by the stiffness. If you get something sensible, then you can replace this simple model with the lookup table and the projection instead of the Von Mises rescaling. I hope this is helpful and not too simple to start with!

**Reply by mathl79** [reply]
too  simple.... not really.
Thanks for your opinion. 
As I often tend to overcomplicate things, i will try to follow your approach first, which sounds very  reasonable to me.

**Reply by mathl79** [reply]
Hi,

could you please check my pull draft #560 where i implemented the discussed plasticity idea and stress coloring?

Do you maybe have an idea how to setup a proper unit test for the plasticity model to check for reasonable outputs ?

I did not succeed to separate the implementations for the plasticity and the stress coloring into different commits. 
Git and contribution in general is new to me.
What is the best practice here ?

**Reply by quagla** [reply]
Thank you so much! Can't wait to play with your changes as soon as I have time. We will also discuss internally and will get back to you.
